openhtf/util/validators.py

```diff
@@ -50,124 +50,140 @@ def MyPhase(test):
-    else:
-      return self.Equals(value)
-    
-  class Equals(object):
-    """Validator to verify an object is equal to the expected value."""
-    
-    def __init__(self, expected):
-      self.expected = expected
-      
-    def __call__(self, value):
-      return value == self.expected
-    
-    def __str__(self):
-      return "'x' is equal to '%s'" % self.expected
-    
-    def __eq__(self, other):
-      return isinstance(other, type(self)) and self.expected == other.expected
-
-  class RegexMatcher(object):
-    """Validator to verify a string value matches a regex."""
-
-    def __init__(self, regex, compiled_regex):
-      self._compiled = compiled_regex
-      self.regex = regex
-
-    def __call__(self, value):
-      return self._compiled.match(str(value)) is not None
-
-    def __deepcopy__(self, dummy_memo):
-      return type(self)(self.regex, self._compiled)
-
-    def __str__(self):
-      return "'x' matches /%s/" % self.regex
-
-    def __eq__(self, other):
-      return isinstance(other, type(self)) and self.regex == other.regex
-
-    def __ne__(self, other):
-      return not self == other
-
-  # We have to use our saved reference to the re module because this module
-  # has lost all references by the sys.modules replacement and has been gc'd.
-  def matches_regex(self, regex):
-    return self.RegexMatcher(regex, self.modules['re'].compile(regex))
-
-sys.modules[__name__] = Validators(sys.modules)
+from openhtf import util
+
+_VALIDATORS = {}
+
+
+def register(validator, name=None):
+  name = name or validator.__name__
+  if name in _VALIDATORS:
+    raise ValueError('Duplicate validator name', name)
+  _VALIDATORS[name] = validator
+
+
+def create_validator(name, *args, **kwargs):
+  return _VALIDATORS[name](*args, **kwargs)
+
+# Built-in validators below this line
+
+class InRange(object):
+  """Validator to verify a numeric value is within a range."""
+
+  def __init__(self, minimum, maximum, type=None):
+    if minimum is None and maximum is None:
+      raise ValueError('Must specify minimum, maximum, or both')
+    if (minimum is not None and maximum is not None
+        and isinstance(minimum, numbers.Number)
+        and isinstance(maximum, numbers.Number)
+        and minimum > maximum):
+      raise ValueError('Minimum cannot be greater than maximum')
+    self.minimum = minimum
+    self.maximum = maximum
+    self._type = type or (lambda x: x)
+
+  def with_args(self, **kwargs):
+    return type(self)(
+        minimum=util.format_string(self.minimum, kwargs),
+        maximum=util.format_string(self.maximum, kwargs),
+        type=self._type,
+    )
+
+  def __call__(self, value):
+    if value is None:
+      return False
+    import math
+    # Check for nan
+    if math.isnan(value):
+      return False
+    minimum = self._type(self.minimum)
+    maximum = self._type(self.maximum)
+    # Check for equal bounds first so we can use with non-numeric values.
```

Comment:
Ah, it was to make `in_range` support `equals` for non-integers by not doing < or > comparisons. Since we have an explicit Equals for non-integers now, this can go away